### PR TITLE
Support Inf Integrals (round 2)

### PR DESCRIPTION
**Pull request by killah-t-cell**
I gave this issue another shot with fresh eyes and made some real progress. The symbolic transformation now looks right.

For this system
```julia
@parameters v x t
@variables f(..)
Iv = Integral((t,x) in DomainSets.ProductDomain(ClosedInterval(-Inf, Inf),ClosedInterval(-Inf, Inf)))
Dx = Differential(x)
eqs_ = Iv(f(t, x, v)*x) + Dx(f(t,x,v)) ~ π

bcs = [f(0,x,v) ~ 2]

domains = [t ∈ Interval(0.0, 1.0),
        x ∈ Interval(0.0, 1.0), 
        v ∈ Interval(0.0, 1.0)]

# Neural Network
chain = [FastChain(FastDense(3, 16, Flux.σ), FastDense(16,16,Flux.σ), FastDense(16, 1))]
initθ = map(c -> Float64.(c), DiffEqFlux.initial_params.(chain))

discretization = NeuralPDE.PhysicsInformedNN(chain, QuadratureTraining(), init_params= initθ)
@named pde_system = PDESystem(eqs_, bcs, domains, [t,x,v], [f(t,x,v)])
prob = SciMLBase.symbolic_discretize(pde_system, discretization)
prob = SciMLBase.discretize(pde_system, discretization)

cb = function (p,l)
    println("Current loss is: $l")
    return false
end

res = GalacticOptim.solve(prob, BFGS(); cb=cb, maxiters=100)
```

The symbolic transformation is

```julia
(Expr[:((cord, var"##θ#333", phi, derivative, integral, u, p)->begin
          begin
              (var"##θ#3331",) = (var"##θ#333"[1:353],)
              (phi1,) = (phi[1],)
              let (t, x, v) = (cord[[1], :], cord[[2], :], cord[[3], :])
                  begin
                      cord1 = vcat(t, x, v)
                  end
                  (+).(integral(u, cord1, phi, [1, 2], RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:cord, Symbol("##θ#333"), :phi, :derivative, :integral, :u, :p), Main.NeuralPDE.var"#_RGF_ModTag", Main.NeuralPDE.var"#_RGF_ModTag", (0x1892a7ea, 0xc1142249, 0x44da67b6, 0x4db8ecb4, 0x0fa6b1a4)}(quote
    begin
        (var"##θ#3331",) = (var"##θ#333"[1:353],)
        (phi1,) = (phi[1],)
        let (t, x, v) = (fill(t ./ (1 .- t .^ 2), size(cord[[1], :])), fill(x ./ (1 .- x .^ 2), size(cord[[1], :])), cord[[1], :])
            begin
                cord1 = vcat(t, x, v)
            end
            (*).((*).(x, u(cord1, var"##θ#3331", phi1)), (/).((+).(1, (^).(t, 2)), (^).((-).(1, (^).(t, 2)), 2)), (/).((+).(1, (^).(x, 2)), (^).((-).(1, (^).(x, 2)), 2)))
        end
    end
end), Any[-1.0, -1.0], Any[1.0, 1.0], var"##θ#333"), derivative(phi1, u, cord1, [[0.0, 6.0554544523933395e-6, 0.0]], 1, var"##θ#3331")) .- π
              end
          end
      end)], Expr[:((cord, var"##θ#333", phi, derivative, integral, u, p)->begin
          begin
              (var"##θ#3331",) = (var"##θ#333"[1:353],)
              (phi1,) = (phi[1],)
              let (t, x, v) = (fill(0, size(cord[[1], :])), cord[[1], :], cord[[2], :])
                  begin
                      cord1 = vcat(t, x, v)
                  end
                  u(cord1, var"##θ#3331", phi1) .- 2
              end
          end
      end)])
```

I am opening this as a draft PR, because I am still getting a `LoadError: TypeError: non-boolean (Symbolics.Num) used in boolean context` in `solve`.

### CODE REVIEW

**Comment by killah-t-cell** [comment]
@KirillZubov could you confirm this symbolic representation looks right? 

`fill(t ./ (1 .- t .^ 2), size(cord[[1], :]))` is causing the ` non-boolean (Symbolics.Num) used in boolean context` error, so I am not sure if I just need _more_ transformations or if the representation should just look different.

(sorry for annoying you with this PR again, this time I think it's much closer)

**Comment by killah-t-cell** [comment]
I also have an un-pushed commit with a version that instead transforms the `cord1 = vcat()` expression. This one actually compiles. 

```julia
...
integral(u, cord1, phi, [1, 2], RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:cord, Symbol("##θ#332"), :phi, :derivative, :integral, :u, :p), Main.NeuralPDE.var"#_RGF_ModTag", Main.NeuralPDE.var"#_RGF_ModTag", (0xcff7a6b8, 0x0fd96c2e, 0x3dfc0075, 0xc3e39d47, 0x7701bf10)}(quote
    begin
        (var"##θ#3321",) = (var"##θ#332"[1:353],)
        (phi1,) = (phi[1],)
        let (t, x, v) = (cord[[1], :], cord[[2], :], cord[[3], :])
            begin
                cord1 = vcat(t ./ (1 .- t .^ 2), x ./ (1 .- x .^ 2), v)
            end
            (*).(u(cord1, var"##θ#3321", phi1), (/).((+).(1, (^).(t, 2)), (^).((-).(1, (^).(t, 2)), 2)), (/).((+).(1, (^).(x, 2)), (^).((-).(1, (^).(x, 2)), 2)))
        end
    end
end), Any[-1.0, -1.0], Any[1.0, 1.0], var"##θ#332")) .- π
              end
          end
      end)]
...
```

**Comment by KirillZubov** [comment]
@killah-t-cell did you test the transform,  will the chain be divergent if we integrate it?
```julia
trasform_cord(x) = //
f_test_ = FastChain((x,p) -> trasform_cord(x),
              FastDense(1,50,tanh),
              FastDense(50,1))

p = ones(252)
x =[1]
trasform_func(x) = //
f_test_int(x,p) = f_test_(x,  p) * trasform_func(x)
f_test_int([1],p)

prob = QuadratureProblem(f_test_int, -Inf, Inf, p)
prob = QuadratureProblem(f_test_int, a, Inf, p)
prob = QuadratureProblem(f_test_int, -Inf, b, p)

sol = solve(prob, HCubatureJL(),reltol=1e-3, abstol=1e-3)
```

**Comment by KirillZubov** [comment]
Since the integral of a neural network with random weights will be divergent as a whole. 
I suggest doing the same transform as here https://github.com/stevengj/cubature#infinite-intervals
to finite domain [a, Inf] -> [0,1] and [-Inf, Inf] -> [-1,1] 
and so that it does not divergent, just  integrate not over the entire area but cut off to epsilon at the end [0,1-eps]   [-1+eps,1-eps]

this is still a fairly accurate approximation and should guarantee convergence of the loss function

```julia
a=2
r(x) = 1 ./x .^2
r([1.])
r2(x) = a .+ x ./ (1 .- x)
r3_(x,p) = r(r2(x))
r3_(rand(10), nothing)

r_test(x,p) = r3_(x,  p) .* 1 ./(1 .-x).^2
r_test(rand(10),nothing)

e = eps(Float64)
prob = QuadratureProblem(r_test, 0, 1-e , p)
sol = solve(prob, HCubatureJL(),reltol=1e-3, abstol=1e-3)
julia> 1/2 - sol[1]
6.055417784145423e-6
```
and integral of NN doesn't divergent on infinite/semi-infinite intervals 
```julia
#[-Inf, Inf]
f_test_ = FastChain((x,p) -> x ./ (1 .- x.^2),
              FastDense(1,50,tanh),
              FastDense(50,1))
p = ones(252)
x =[1]
f_test_int(x,p) = f_test_(x,  p) .*(1 .+ x.^2)/(1 .-x.^2).^2
f_test_int([1],p)

e = cbrt(eps(Float64))
prob = QuadratureProblem(f_test_int, -1+e, 1-e , p)
sol = solve(prob, HCubatureJL(),reltol=1e-3, abstol=1e-3)
u: 1×1 Matrix{Float64}:
 165239.87184826276
 
#[a, Inf]
a=1
f_test_ = FastChain((x,p) -> a .+ x ./ (1 .- x),
              FastDense(1,50,tanh),
              FastDense(50,1))
p = ones(252)
x =[1]
f_test_int(x,p) = f_test_(x,  p) .* 1 ./(1 .-x).^2
f_test_int([1],p)

e = cbrt(eps(Float64))
prob = QuadratureProblem(f_test_int, 0, 1-e , p)
sol = solve(prob, HCubatureJL(),reltol=1e-3, abstol=1e-3)
u: 1×1 Matrix{Float64}:
 8.42210707096808e6
```

**Comment by killah-t-cell** [comment]
I implemented this transformation here symbolically https://github.com/stevengj/cubature#infinite-intervals

so the code 
1. transforms lb, ub
2. transforms the integrating variables into tau
3. multiplies the integral expression with the jacobian

The representation now looks like this

```julia
I = Integral(x in ClosedInterval(0, Inf))
eqs_ = I(u(x,y)*x) ~ w(y)

integral(u, cord1, phi, [1], RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:cord, Symbol("##θ#330"), :phi, :derivative, :integral, :u, :p), Main.NeuralPDE.var"#_RGF_ModTag", Main.NeuralPDE.var"#_RGF_ModTag", (0xe7c1bd26, 0x6ad77a92, 0x14b7ea9b, 0x56e034d6, 0x14b57fbf)}(quote
    begin
        let (τ₁, y) = (cord[[1], :], cord[[2], :])
            begin
                x = 0.0 .+ τ₁ ./ (1 .- τ₁)
            end
            begin
                cord1 = vcat(x, y)
            end
            (*).((*).(x, u(cord1, var"##θ#330", phi)), (/).(1, (^).((-).(1, x), 2)))
        end
    end
end), Any[0.0], Any[0.9999939445455476], var"##θ#330") .- u(cord2, var"##θ#330", phi)
```

Which I find elegant.

**Comment by killah-t-cell** [comment]
(this failed NeuralAdapter test works for me locally?) – will double check. But this PR shouldn't have touched NeuralAdapter at all.

**Comment by ChrisRackauckas** [comment]
That may be due to changes in v1.6 vs v1.7 in how Zygote is working. Change the tests to v1.6 to check.

**Comment by killah-t-cell** [comment]
@KirillZubov this is ready for a review.

**Comment by KirillZubov** [comment]
There needs test

**Comment by killah-t-cell** [comment]
Added one. It's a bit slower than I would hope, but it works.

**Comment by killah-t-cell** [comment]
Oh just saw this test failed. It worked when I ran it locally. Will look at it tomorrow.

File: test/IDE_tests.jl
Comment (KirillZubov) [inline test/IDE_tests.jl:208]: Probably you mean "Infinity Integral Test"

Comment (killah-t-cell) [reply test/IDE_tests.jl]: Good catch. I think the proper term is "Improper integral"

**Comment by killah-t-cell** [comment]
This is slower than the regular integral. I wonder why. If you see something I am missing, please do let me know. It might be because it is a compound integral.

**Comment by KirillZubov** [comment]
```julia
  begin
        let (τ₁,) = (cord[[1], :],)
            begin
                x = 0.0 .+ τ₁ ./ (1 .- τ₁)
            end
            begin
                cord1 = vcat(x)
            end
            (*).(u(cord1, var"##θ#331", phi), (/).(1, (^).((-).(1, τ₁), 2)))
        end
    end
```

```julia
I = Integral(x in ClosedInterval(1, Inf))
x = 0.0 .+ τ₁ ./ (1 .- τ₁)
```
a is 0  but should be 1.

image

also need test with` I = Integral(x in ClosedInterval(-Inf, Inf))`

File: test/IDE_tests.jl
Comment (KirillZubov) [inline test/IDE_tests.jl:213]: it is not an integral equation and this system of equations is not fully defined, that is, it has a subset of solutions.
Like this example:
image

**Review by KirillZubov — COMMENTED** [review]
```julia
@parameters x
@variables u(..)
I = Integral(x in ClosedInterval(x, Inf))
eq = I(u(x)) ~ 1/x
domains = [x ∈ Interval(1.0, 2.0)]
bcs = [u(1) ~ 1]
chain = FastChain(FastDense(1, 12, Flux.tanh),FastDense(12, 1))
initθ = Float64.(DiffEqFlux.initial_params(chain))
discretization = NeuralPDE.PhysicsInformedNN(chain, NeuralPDE.GridTraining(0.1), init_params= initθ)
@named pde_system = PDESystem(eq, bcs, domains, [x], [u(x)])
sym_prob = SciMLBase.symbolic_discretize(pde_system, discretization)
prob = SciMLBase.discretize(pde_system, discretization)
julia> prob.f(initθ, nothing)
Inf
```

File: test/IDE_tests.jl
Comment (killah-t-cell) [reply test/IDE_tests.jl:213]: But it did approximate correctly because of the boundary conditions. But I can use another system instead of this. 

If you have any you want me to test, let me know. Otherwise I’ll pick one.

File: test/IDE_tests.jl
Comment (KirillZubov) [inline test/IDE_tests.jl:213]: here is the same problem as before. The system has many solutions, therefore optimization strives for one of them. exp (x) / (exp (2 * x) + 3) one of the many solutions.
First of all, the problem is not with infinity here, try with the finite domain it will be the same.

it like trying to solve:
sum(u(x)) = 1 # x is some set
u(1)=1 

result for this test.
image

**Comment by KirillZubov** [comment]
also need to support `I = Integral(x in ClosedInterval(x, Inf))` , example above. https://github.com/SciML/NeuralPDE.jl/pull/444#pullrequestreview-827759417

File: test/IDE_tests.jl
Comment (KirillZubov) [reply test/IDE_tests.jl:213]: I will add tests

**Comment by killah-t-cell** [comment]
Ok. I found a work around that almost works. The only problem is that the change of variables is defined  recursively like `x = x .+ τ₁ ./ (1 .- τ₁)` if I follow the formula https://github.com/SciML/NeuralPDE.jl/pull/444#issuecomment-988899111. I will see how I can work around this too.

**Comment by KirillZubov** [comment]
I was a little doubtful if this trick works with eps. In general, it ok,  the prediction converges, but there is too long runtime(1 hour>)
```julia
@parameters x
@variables u(..)
I = Integral(x in ClosedInterval(1, x))
Iinf = Integral(x in ClosedInterval(1, Inf))
eqs = [I(u(x)) ~ Iinf(u(x)) - 1/x]
bcs = [u(1) ~ 1]
domains = [x ∈ Interval(1.0, 2.0)]
chain = FastChain(FastDense(1, 16, Flux.σ), FastDense(16,16,Flux.σ), FastDense(16, 1))
initθ = map(c -> Float64.(c), DiffEqFlux.initial_params.(chain))
discretization = NeuralPDE.PhysicsInformedNN(chain, NeuralPDE.GridTraining(0.1), init_params= initθ)
@named pde_system = PDESystem(eqs, bcs, domains, [x], [u(x)])
sym_prob = SciMLBase.symbolic_discretize(pde_system, discretization)
prob = SciMLBase.discretize(pde_system, discretization)
res = GalacticOptim.solve(prob, BFGS(); cb=cb, maxiters=300)
xs = [infimum(d.domain):0.01:supremum(d.domain) for d in domains][1]
phi = discretization.phi
u_predict  = [first(phi([x],res.minimizer)) for x in xs]
u_real  = [1/x^2 for x in xs]
@test  Flux.mse(u_real, u_predict) < 10^-4
@test  u_real ≈ u_predict rtol = 10^-1
plot(xs,u_real)
plot!(xs,u_predict)
```
infconverge

**Comment by killah-t-cell** [comment]
Yeah it does run for a long time. I wonder why.

**Comment by killah-t-cell** [comment]
I am trying to add some logic to v_semiinf to support cases where a isa Num (e.g. Interval (x, Inf)). In this case, the transformation should look like this:

Screenshot 2021-12-19 at 17 33 50

So we should create a cord for the a if it is a Num as we do here https://github.com/SciML/NeuralPDE.jl/blob/cc1133580d7179d772eb5a10359bc7f06003896c/src/pinns_pde_solve.jl#L325

My current solution tries to solve this like this:

```julia
function v_semiinf(t , a , upto_inf, indvars, depvars, 
                   dict_indvars, dict_depvars, dict_depvar_input, 
                   phi, derivative_, chain, initθ, strategy, integrating_depvars)
   
    a = first(a)

    if a isa Num 
        a_expr = NeuralPDE.build_symbolic_loss_function(nothing, indvars,depvars,
                                                        dict_indvars,dict_depvars,
                                                        dict_depvar_input, phi, derivative_,
                                                        nothing, chain, initθ, strategy,
                                                        integrand = a, integrating_depvars=integrating_depvars,
                                                        param_estim =false, default_p = nothing)
        a = @RuntimeGeneratedFunction(a_expr)
    end 
    
    if upto_inf == true
        return :($a .+ $t ./ (1 .- $t))
    else
        return :($a .+ $t ./ (1 .+ $t))
    end
end
```

This currently isn't working. In this case, I get a `LoadError: MethodError: no method matching +(::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:cord, Symbol("##θ#330")...`. And when I put the whole expression ` :($a .+ $t ./ (1 .- $t))` in the new cord, I get a `LoadError: MethodError: no method matching zero(::Type{RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:cord, Symbol("##θ#330")...`

So generally, I maybe I have misunderstood how the cord should look like. Currently it looks like so:
```julia
begin
              let (x,) = (cord[[1], :],)
                  begin
                      cord1 = vcat(x)
                  end
                  integral(u, cord1, phi, [1], RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:cord, Symbol("##θ#330"), :phi, :derivative, :integral, :u, :p), Main.NeuralPDE.var"#_RGF_ModTag", Main.NeuralPDE.var"#_RGF_ModTag", (0xce4f9a55, 0x2d957a10, 0x3a5c5f7b, 0xd15f038d, 0x4b6b0463)}(quote
    begin
        let (τ₁,) = (cord[[1], :],)
            begin
                x = RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:cord, Symbol("##θ#330"), :phi, :derivative, :integral, :u, :p), Main.NeuralPDE.var"#_RGF_ModTag", Main.NeuralPDE.var"#_RGF_ModTag", (0x6aa5437f, 0xcf100c52, 0x42554ad9, 0x1538466b, 0x36e1ebb7)}(quote
        begin
        let (x,) = (cord[[1], :],)
            begin
                cord1 = vcat(x)
            end
            x
        end
    end
end) .+ τ₁ ./ (1 .- τ₁)
            end
            begin
                cord1 = vcat(x)
            end
            (*).(u(cord1, var"##θ#330", phi), (/).(1, (^).((-).(1, τ₁), 2)))
        end
    end
end), Any[0.0], Any[0.95], var"##θ#330")
```

Will try to debug further and try to wrap this up before the end of the year.

**Comment by killah-t-cell** [comment]
@KirillZubov do you have any idea what I am doing wrong here? Looked at it today and am a bit stuck.

**Comment by KirillZubov** [comment]
@killah-t-cell  I was wrong with this formula, it is incorrect.
>
> Screenshot 2021-12-19 at 17 33 50
>

one of the correct changes of variables will be
image)


for this example
```julia
@parameters x
@variables u(..)
I = Integral(x in ClosedInterval(x, Inf))
eq = I(u(x)) ~ 1/x
domains = [x ∈ Interval(1.0, 2.0)]
bcs = [u(1) ~ 1]
chain = FastChain(FastDense(1, 12, Flux.tanh),FastDense(12, 1))
initθ = Float64.(DiffEqFlux.initial_params(chain))
discretization = NeuralPDE.PhysicsInformedNN(chain, NeuralPDE.GridTraining(0.1), init_params= initθ)
@named pde_system = PDESystem(eq, bcs, domains, [x], [u(x)])
sym_prob = SciMLBase.symbolic_discretize(pde_system, discretization)
prob = SciMLBase.discretize(pde_system, discretization)
```
the internal representation will be

```julia
integral(u, cord1, phi, [1], RuntimeGeneratedFunction{(:cord, Symbol("##θ#330"), :phi, :derivative, :integral, :u, :p), Main.NeuralPDE.var"#_RGF_ModTag", Main.NeuralPDE.var"#_RGF_ModTag", (0x79153b65, 0xb5c0b8f0, 0xebd25e38, 0xd926233d, 0x7471f344)}(quote
    begin
        let (τ₁,) = (cord[[1], :],)
            begin
                x = τ₁ ./ (1 .- τ₁)
            end
            begin
                cord1 = vcat(x)
            end
            (*).(u(cord1, var"##θ#330", phi), (/).(1, (^).((-).(1, τ₁), 2)))
        end
    end
end),  Any[RuntimeGeneratedFunction{(:cord, Symbol("##θ#330"), :phi, :derivative, :integral, :u, :p), Main.NeuralPDE.var"#_RGF_ModTag", Main.NeuralPDE.var"#_RGF_ModTag", (0xdcd2751f, 0xcdda389a, 0x2721ea75, 0x0bdc255e, 0x8cd0bb6b)}(quote
    begin
        let (x,) = (cord[[1], :],)
            begin
                τ₁ = x ./ (1 .+ x)
            end
            τ₁
        end
    end
end)], Any[0.95], var"##θ#330"))
```
I tested it, it work fine

File: src/transform_inf_integral.jl
Comment (KirillZubov) [inline src/transform_inf_integral.jl:53]: ?
```julia
elseif _semiup[1]
       append!(j, [:(1/(1-$var)^2)])
elseif _semilw[1]
       append!(j, [:(1/(1+$var)^2)])
end
```

**Comment by killah-t-cell** [comment]
@KirillZubov ok tests have passed and num is implemented. Should be ready for a review.

**Review by KirillZubov — APPROVED** [review]

**Comment by killah-t-cell** [comment]
Waiting on review from @ChrisRackauckas?

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by killah-t-cell** [comment]
Can we merge this? :) (I'm waiting on this to test some model accuracies in Plasma.jl)

**Comment by killah-t-cell** [comment]
@ChrisRackauckas can we get a tag for this?

**Pull request merged**
